File: app/src/main/java/mk/webfactory/template/data/storage/JsonConverter.java
package mk.webfactory.template.data.storage;

import com.google.gson.Gson;
import java.lang.reflect.Type;

public interface JsonConverter {

  String toJson(Object o) throws Exception;

  <T> T fromJson(String json, Type type) throws Exception;

  class GsonConverter implements JsonConverter {

    private final Gson gson;

    public GsonConverter(Gson gson) {
      this.gson = gson;
    }

    @Override public String toJson(Object o) {
      return gson.toJson(o);
    }

    @Override public <T> T fromJson(String json, Type type) {
      return gson.fromJson(json, type);
    }
  }
}

File: app/src/main/java/mk/webfactory/template/data/storage/JsonConverter.kt
package mk.webfactory.template.data.storage

import com.google.gson.Gson
import java.lang.reflect.Type

interface JsonConverter {
    @Throws(Exception::class)
    fun toJson(o: Any): String

    @Throws(Exception::class)
    fun <T> fromJson(json: String, type: Type): T

    class GsonConverter(private val gson: Gson) : JsonConverter {
        override fun toJson(o: Any): String {
            return gson.toJson(o)
        }

        override fun <T> fromJson(json: String, type: Type): T {
            return gson.fromJson(json, type)
        }

    }
}
